FEXHeaderUtils/FEXHeaderUtils/ScopedSignalMask.h

```diff
@@ -43,4 +46,49 @@ namespace FHU {
       uint64_t OriginalMask{};
       std::mutex &Mutex;
   };
+
+  /**
+   * @brief A class that masks signals and locks a shared mutex until it goes out of scope
```

Comment:
> When do I need to worry about signal-reentrancy?
Unknown. Always? Right now we're dealing with it on a case-by-case basis when it becomes a problem.

> When does this class need to be used over ScopedSignalMaskWithMutex?
When you want `std::shared_mutex` and not `std::mutex`

> If I'm adding new code, how do I know that I need to use this class?
> If I'm modifying existing code, how do I know that use of `ScopedSignalMask*` can safely be dropped?

These are open questions, without a clear guideline, and I don't think they can be answered until we have a concrete gameplan for handling signals safely.

Reply:
> These are open questions, without a clear guideline, and I don't think they can be answered until we have a concrete gameplan for handling signals safely.

Looks like it's unrealistic that we're going to come up with a better solution as part of this PR then. Let's work on minimizing the additional complexity needed to fit the existing design.

> When do I need to worry about signal-reentrancy? When does this class need to be used over ScopedSignalMaskWithMutex? What fundamental issue does it address regarding FEX's functionality? If I'm adding new code, how do I know that I need to use this class? How do I know the other one is needed? If I'm modifying existing code, how do I know that use of ScopedSignalMask* can safely be dropped?

Ultimately ScopedSignalMaskWithMutex and ScopedSignalMaskWithSharedMutex are two separate classes that do the same thing: Lock the given object in a matter that's robust against the signal shenanigans we need to deal with.

Why not use a single class instead? By removing the ambiguity, three of my questions above are resolved, leaving just the other three to be addressed with a bigger-picture solution.

This could be implemented using `std::variant` like this (also pulling in my suggestion about avoiding boolean blindness):
```cpp
class ScopedSignalMaskWithMutex {
    ScopedSignalMaskWithMutex(std::shared_mutex &_Mutex, bool shared, uint64_t Mask) : mutex(_Mutex), shared(shared) { ... }

public:
    ScopedSignalMaskWithMutex(std::mutex &_Mutex, uint64_t Mask = ~0ULL) : mutex(_Mutex), shared(shared) { ... }

    ScopedSignalMaskWithMutex WithSharedLock(std::shared_mutex &_Mutex, uint64_t Mask = ~0ULL);
    ScopedSignalMaskWithMutex WithExclusiveLock(std::shared_mutex &_Mutex, uint64_t Mask = ~0ULL);

    ~ScopedSignalMaskWithMutex() {
        if (auto* unshared_mutex = std::get_if<std::mutex>(&mutex)) {
            unshared_mutex->unlock();
        } else {
            auto& shared_mutex = std::get<std::shared_mutex>(mutex);
            if (shared) {
                shared_mutex.unlock_shared();
            } else {
                shared_mutex.unlock();
            }
        }
    }

private:
    std::variant<std::mutex, std::shared_mutex> mutex;
    bool is_shared;
};
```

Reply:
Your enum proposal would also work with this approach, in which case we could just provide two overloaded constructors, which would make usage even more straightforward.

Reply:
(this is now in #1677)